libbeat/publisher/queue/diskqueue/writer_loop.go

```diff
@@ -91,9 +92,13 @@ type writerLoop struct {
 	outputFile *os.File
 
 	currentRetryInterval time.Duration
+
+	// buffer Used to gather write information so there is only one write syscall
+	buffer *bytes.Buffer
 }
 
 func newWriterLoop(logger *logp.Logger, settings Settings) *writerLoop {
+	buffer := &bytes.Buffer{}
```

Comment:
As another minor optimization, are you able to know how big this needs to be before using it? You could make sure it only allocates once with https://pkg.go.dev/bytes#Buffer.Grow

Alternatively we could reuse it between calls with https://pkg.go.dev/bytes#Buffer.Reset but we'll hang on to the largest size buffer ever used indefinitely.

Reply:
The writer loop is only initialized once, so this bytes.Buffer is used for the lifetime of the queue.  Unfortunately when the writer_loop is started no writes have come in yet so we don't know how big to make it.

There is a call to Buffer.Reset(), so we are keeping the largest buffer used for the lifetime of the queue.  It shouldn't be huge, the size of the largest frame, which will only have one event.

Reply:
Thanks! I was mislead by this being in a function where I could see the lifecycle in the diff.